### PR TITLE
Implement smoothen options

### PR DESCRIPTION
**Pull request by Sjoerd1993**
This implements extra options for smoothening. This includes a new filter technique, the Savitzky–Golay filter which allows to do some pretty cool filtering. See example in the screencast, where I'm filtering a noisy sine wave. The original sine wave is has an amplitude of one, with some `np.random` noise added to it. As you can see, with a moving average you won't really get a smooth sine back, but the Savitzky–Golay draws a polynomial using defined window ranges and comes with this much nicer graph.

This all works now, but it probably needs some cleaning and it will not pass linting likely. I'm already running late though, so I wanted to put this up before I start on work.

[Screencast from 2023-11-17 08-35-11.webm](https://github.com/Sjoerd1993/Graphs/assets/68477016/f82b8fbc-8491-452a-8d7b-64db6561d1f0)

### CODE REVIEW

File: data/se.sjoerd.Graphs.gschema.xml [hunk 1]
```diff
@@ -46,6 +51,7 @@
 
   <schema id="se.sjoerd.Graphs" path="/se/sjoerd/Graphs/">
     <child name="actions" schema="se.sjoerd.Graphs.actions"/>
+    <child name="smoothen" schema="se.sjoerd.Graphs.smoothen"/>
```
Comment (cmkohnen) [inline data/se.sjoerd.Graphs.gschema.xml:54]: This is redundant with it also being present in actions

File: src/actions.py [hunk 2]
```diff
@@ -153,3 +165,38 @@ def delete_selected_action(_action, _target, self):
     names = ", ".join(item.get_name() for item in items)
     self.get_data().delete_items(items)
     self.get_window().add_toast_string(_("Deleted {}").format(names))
+
+
+@Gtk.Template(resource_path="/se/sjoerd/Graphs/ui/smoothen_settings.ui")
```
Comment (cmkohnen) [inline src/actions.py:170]: This can probably moved to vala in a separate pr

File: src/actions.py [hunk 3]
```diff
@@ -18,8 +20,14 @@ def perform_operation(_action, target, self):
...
-    if operation in ("center", ):
+    if operation in ("center", "smoothen"):
         args = [self.get_settings("actions").get_enum(operation)]
+    if operation == "smoothen":
+        params = {}
+        settings = self.get_settings("actions").get_child("smoothen")
+        for setting in settings:
+            params[setting] = int(settings.get_int(setting))
+        args += [params]
     if operation == "shift":
```
Comment (cmkohnen) [inline src/actions.py:31]:
```suggestion
    elif operation == "shift":
```

File: src/application.py [hunk 4]
```diff
@@ -102,6 +103,10 @@ def __init__(self, application_id, **kwargs):
         center_action = settings.get_child("actions").create_action("center")
```
Comment (cmkohnen) [inline src/application.py:103]: As more actions are probably implemented using this way, a loop structure would be useful here. Can be part of a separate pr however

**Review by cmkohnen — CHANGES_REQUESTED** [review]

File: src/actions.py (hunk 2 above)
Comment (Sjoerd1993) [reply src/actions.py]: I moved most of it to Vala, but couldn't get the reset button to work. So that's still in the python code for now.

File: src/application.py [hunk 5]
```diff
@@ -99,8 +100,9 @@ def __init__(self, application_id, **kwargs):
         operation_action.connect("activate", actions.perform_operation, self)
         self.add_action(operation_action)
 
-        center_action = settings.get_child("actions").create_action("center")
-        self.add_action(center_action)
+        for action_key in ["center", "smoothen"]:
+            action = settings.get_child("actions").create_action(action_key)
```
Comment (cmkohnen) [inline src/application.py:104]: I'd outline the get_child("actions") to be outside the loop

File: src/smoothen_settings.vala [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        [GtkChild]
+        public unowned Button reset_button { get; }
+
+        public SmoothenWindow (Application application) {
+            Object (
+                application: application,
+                transient_for: application.window
+            );
+            Tools.bind_settings_to_widgets (
+                application.settings.get_child("actions").get_child("smoothen"), this
```
Comment (cmkohnen) [inline src/smoothen_settings.vala:26]:
```suggestion
                application.settings.get_child ("actions").get_child ("smoothen"), this
```

File: src/actions.py [hunk 7]
```diff
@@ -123,6 +131,16 @@ def save_project_action(_action, _target, self):
     ui.save_project_dialog(self)
 
 
+def smoothen_settings_action(_action, _target, self):
+    def _on_reset(_button, self):
+        params = self.get_settings("actions").get_child("smoothen")
+        for key in params.list_keys():
+            params.reset(key)
```
Comment (cmkohnen) [inline src/actions.py:138]: SInce this shares functionality with the import settings, perhaps there could be a way to move it to utilities? or even to vala, that way we can make direct use of builder callbacks

**Review by cmkohnen — CHANGES_REQUESTED** [review]
some minor stuff still, otherwise this looks good

**Review by cmkohnen — APPROVED** [review]
